File: BE/CpuA64/symbolic.py
"""
This module contains code for (un-)symbolizing the a64 ISA operands
"""
from typing import List, Tuple
import struct

from BE.Elf import enum_tab
from BE.CpuA64 import opcode_tab as a64


def _FloatTo64BitRepresentation(num: float) -> int:
    b = struct.pack('<d', num)
    assert len(b) == 8
    return int.from_bytes(b, "little")


def _Flt64FromBits(data: int) -> float:
    return struct.unpack('<d', int.to_bytes(data, 8, "little"))[0]


def SymbolizeOperand(ok: a64.OK, data: int) -> str:
    t = a64.FIELD_DETAILS.get(ok)
    assert t is not None, f"NYI: {ok}"
    data = a64.DecodeOperand(ok, data)
    if t.kind == a64.FK.LIST:
        return t.names[data]
    elif t.kind == a64.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = _Flt64FromBits(data)
        return f"{data:g}"
    elif t.kind == a64.FK.INT_SIGNED:
        # we only care about the signed aspect
        if data >= 1 << 63:
            data -= (1 << 64)
        return str(data)
    elif t.kind == a64.FK.INT_HEX or t.kind == a64.FK.INT_HEX_CUSTOM:
        return hex(data)
    else:
        return str(data)


def UnsymbolizeOperand(ok: a64.OK, op: str) -> int:
    """
    Converts a string into and int suitable for the provided `ok`
    E.g.
    #66 -> 65
    x0 -> 0
    asr -> 2
    """
    t = a64.FIELD_DETAILS.get(ok)
    assert t is not None, f"NYI: {ok}"

    if t.kind == a64.FK.LIST:
        # assert op in t, f"{op} not in {t}"
        data = t.names.index(op)
    elif t.kind == a64.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = _FloatTo64BitRepresentation(float(op))
    else:
        data = int(op, 0)  # skip "#", must handle "0x" prefix
        # note we intentionally allow negative numbers here
    return a64.EncodeOperand(ok, data)


_RELOC_KIND_MAP = {
    # these relocations imply that the symbol is local
    "jump26": (enum_tab.RELOC_TYPE_AARCH64.JUMP26, True),
    "condbr19": (enum_tab.RELOC_TYPE_AARCH64.CONDBR19, True),
    #
    "call26": (enum_tab.RELOC_TYPE_AARCH64.CALL26, False),
    "abs32": (enum_tab.RELOC_TYPE_AARCH64.ABS32, False),
    "abs64": (enum_tab.RELOC_TYPE_AARCH64.ABS64, False),
    "adr_prel_pg_hi21": (enum_tab.RELOC_TYPE_AARCH64.ADR_PREL_PG_HI21, False),
    "add_abs_lo12_nc": (enum_tab.RELOC_TYPE_AARCH64.ADD_ABS_LO12_NC, False),
    #
    "loc_call26": (enum_tab.RELOC_TYPE_AARCH64.CALL26, True),
    "loc_abs32": (enum_tab.RELOC_TYPE_AARCH64.ABS32, True),
    "loc_abs64": (enum_tab.RELOC_TYPE_AARCH64.ABS64, True),
    "loc_adr_prel_pg_hi21": (enum_tab.RELOC_TYPE_AARCH64.ADR_PREL_PG_HI21, True),
    "loc_add_abs_lo12_nc": (enum_tab.RELOC_TYPE_AARCH64.ADD_ABS_LO12_NC, True),
}

_RELOC_OK = {
    a64.OK.SIMM_PCREL_0_25,  # RELOC_TYPE_AARCH64.JUMP26
    a64.OK.SIMM_PCREL_5_23,
    a64.OK.SIMM_PCREL_5_23_29_30,
    a64.OK.IMM_SHIFTED_10_21_22,  # RELOC_TYPE_AARCH64.ADD_ABS_LO12_NC
}


def _EmitReloc(ins: a64.Ins, pos: int) -> str:
    if ins.reloc_kind == enum_tab.RELOC_TYPE_AARCH64.JUMP26:
        assert ins.is_local_sym, f"expected local symbol"
        return f"expr:jump26:{ins.reloc_symbol}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_AARCH64.ADR_PREL_PG_HI21:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}adr_prel_pg_hi21:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_AARCH64.ADD_ABS_LO12_NC:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}add_abs_lo12_nc:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_AARCH64.CALL26:
        return f"expr:call26:{ins.reloc_symbol}"
    if ins.reloc_kind == enum_tab.RELOC_TYPE_AARCH64.CONDBR19:
        assert ins.is_local_sym, f"expected local symbol"
        return f"expr:condbr19:{ins.reloc_symbol}"
    else:
        assert False


def InsSymbolize(ins: a64.Ins) -> Tuple[str, List[str]]:
    """Convert all the operands in an arm.Ins to strings including relocs
    """
    ops = []
    for pos, (ok, value) in enumerate(zip(ins.opcode.fields, ins.operands)):
        if ins.has_reloc() and ins.reloc_pos == pos:
            assert ok in _RELOC_OK
            ops.append(_EmitReloc(ins, pos))
        else:
            ops.append(SymbolizeOperand(ok, value))

    return ins.opcode.NameForEnum(), ops


def InsFromSymbolized(mnemonic: str, ops_str: List[str]) -> a64.Ins:
    opcode = a64.Opcode.name_to_opcode[mnemonic]
    ins = a64.Ins(opcode)
    for pos, (t, ok) in enumerate(zip(ops_str, opcode.fields)):
        if t.startswith("expr:"):
            # expr strings have the form expr:<rel-kind>:<symbol>:<addend>, e.g.:
            #   expr:movw_abs_nc:string_pointers:5
            #   expr:call:putchar
            rel_token = t.split(":")
            assert rel_token[1] in _RELOC_KIND_MAP, f"unknown reloc kind {rel_token}"
            ins.set_reloc(*_RELOC_KIND_MAP[rel_token[1]], pos, rel_token[2])
            addend = 0 if len(rel_token) == 3 else int(rel_token[3], 0)
            ins.operands.append(addend)
        else:
            ins.operands.append(UnsymbolizeOperand(ok, t))
    return ins

File: BE/CpuA64/symbolic.cc

#include "BE/CpuA64/symbolic.h"

#include <cstring>

#include "Util/assert.h"
#include "Util/parse.h"

namespace cwerg::a64 {
using namespace cwerg;

char* strappend(char* dst, std::string_view src) {
  std::memcpy(dst, src.data(), src.size());
  dst[src.size()] = 0;
  return dst + src.size();
}

char* strappenddec(char* dst, int64_t n) {
  if (n >= 0) {
    return dst + ToDecString(n, dst).size();
  } else {
    *dst++ = '-';
    return dst + ToDecString(-n, dst).size();
  }
}

char* strappendhex(char* dst, int64_t n) {
  dst = strappend(dst, "0x");
  return dst + ToHexString(n, dst).size();
}

char* strappendflt(char* dst, double d) {
  return dst + ToFltString(d, dst).size();
}

double Flt64FromBits(uint64_t i) {
  union {
    uint64_t i;
    double d;
  } u = {i};
  return u.d;
}

uint64_t Flt64ToBits(double d) {
  union {
    double d;
    uint64_t i;
  } u = {d};
  return u.i;
}

char* SymbolizeOperand(char* buf, uint32_t data, OK ok) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  const uint64_t val = DecodeOperand(ok, data);
  switch (fi.kind) {
    default:
      ASSERT(false, "");
      return nullptr;
    case FK::LIST:
      return strappend(buf, fi.names[val]);
    case FK::FLT_CUSTOM: {
      return strappendflt(buf, Flt64FromBits(val));
    }
    case FK::INT_SIGNED:
    case FK::INT:  // TODO: add unsigned version
      return strappenddec(buf, (int64_t)val);
    case FK::INT_HEX_CUSTOM:
    case FK::INT_HEX:
      return strappendhex(buf, (int64_t)val);
  }
}

void SymbolizeReloc(char* cp, const Ins& ins, uint32_t addend) {
  cp = strappend(cp, "expr:");
  switch (ins.reloc_kind) {
    case elf::RELOC_TYPE_AARCH64::JUMP26:
      ASSERT(ins.is_local_sym, "");
      cp = strappend(cp, "jump26:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_AARCH64::ADR_PREL_PG_HI21:
      cp = strappend(
          cp, ins.is_local_sym ? "loc_adr_prel_pg_hi21:" : "adr_prel_pg_hi21:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_AARCH64::ADD_ABS_LO12_NC:
      cp = strappend(
          cp, ins.is_local_sym ? "loc_add_abs_lo12_nc:" : "add_abs_lo12_nc:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_AARCH64::CALL26:
      cp = strappend(cp, "call26:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_AARCH64::CONDBR19:
      ASSERT(ins.is_local_sym, "");
      cp = strappend(cp, "condbr19:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    default:
      ASSERT(false, "");
  }
  if (addend != 0) {
    *cp++ = ':';
    cp = strappenddec(cp, addend);
  }
}

std::string_view InsSymbolize(const a64::Ins& ins,
                              std::vector<std::string>* ops) {
  char buffer[128];
  for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
    if (ins.has_reloc() && i == ins.reloc_pos) {
      SymbolizeReloc(buffer, ins, ins.operands[i]);
    } else {
      SymbolizeOperand(buffer, ins.operands[i], ins.opcode->fields[i]);
    }
    ops->emplace_back(buffer);
  }
  return ins.opcode->name;
}

uint32_t UnsymbolizeOperand(OK ok, std::string_view op) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  uint64_t val;
  switch (fi.kind) {
    default:
      ASSERT(false, "");
      return kEncodeFailure;
    case FK::LIST:
      for (val = 0; val < fi.num_names; val++) {
        if (fi.names[val] == op) break;
      }
      if (val == fi.num_names) return kEncodeFailure;
      break;
    case FK::FLT_CUSTOM: {
      auto maybe = ParseReal(op);
      if (!maybe) return kEncodeFailure;
      val = Flt64ToBits(maybe.value());
      break;
    }
    case FK::INT_SIGNED: {
      auto maybe = ParseInt<int64_t>(op);
      if (!maybe) return kEncodeFailure;
      val = maybe.value();
      break;
    }
    case FK::INT:  // TODO: add unsigned version
    case FK::INT_HEX_CUSTOM:
    case FK::INT_HEX: {
      auto maybe = ParseInt<uint64_t>(op);
      if (!maybe) return kEncodeFailure;
      val = maybe.value();
      break;
    }
  }

  return EncodeOperand(ok, val);
}

const constexpr struct {
  std::string_view expr_kind;
  elf::RELOC_TYPE_AARCH64 reloc_kind;
  bool local;
} kExprToReloc[] = {
    // order by frequency
    {"jump26", elf::RELOC_TYPE_AARCH64::JUMP26, true},
    {"condbr19", elf::RELOC_TYPE_AARCH64::CONDBR19, true},
    {"call26", elf::RELOC_TYPE_AARCH64::CALL26, false},
    {"adr_prel_pg_hi21", elf::RELOC_TYPE_AARCH64::ADR_PREL_PG_HI21, false},
    {"add_abs_lo12_nc", elf::RELOC_TYPE_AARCH64::ADD_ABS_LO12_NC, false},
    {"loc_adr_prel_pg_hi21", elf::RELOC_TYPE_AARCH64::ADR_PREL_PG_HI21, true},
    {"loc_add_abs_lo12_nc", elf::RELOC_TYPE_AARCH64::ADD_ABS_LO12_NC, true},
    {"abs32", elf::RELOC_TYPE_AARCH64::ABS32, false},
    {"abs64", elf::RELOC_TYPE_AARCH64::ABS64, false},
};

bool HandleRelocation(std::string_view expr, unsigned pos, Ins* ins) {
  auto expr_op = ParseExpressionOp(expr);
  if (!expr_op) return false;
  ins->operands[pos] = expr_op->offset;
  for (const auto& x : kExprToReloc) {
    if (x.expr_kind == expr_op->reloc_name) {
      ins->set_reloc(x.reloc_kind, x.local, pos, expr_op->symbol_name);
      return true;
    }
  }
  return false;
}

bool InsFromSymbolized(const std::vector<std::string_view>& token, Ins* ins) {
  ins->opcode = FindOpcodeForMnemonic(token[0]);
  if (ins->opcode == nullptr) {
    std::cerr << "unknown opcode " << token[0] << "\n";
    return false;
  }
  ASSERT(token.size() - 1 == ins->opcode->num_fields,
         "bad number of token " << token.size() << " expected "
                                << ins->opcode->num_fields + 1 << " for "
                                << token[0]);
  for (unsigned i = 1; i < token.size(); ++i) {
    if (token[i].substr(0, 5) == "expr:") {
      if (!HandleRelocation(token[i].substr(5), i - 1, ins)) {
        std::cerr << "malformed relocation expression " << token[i] << "\n";
        return false;
      }
    } else {
      const OK ok = ins->opcode->fields[i - 1];
      auto val = UnsymbolizeOperand(ok, token[i]);
      if (val == kEncodeFailure) {
        std::cerr << "cannot parse " << token[i] << " for ok [" << int(ok)
                  << "]\n";
        return false;
      }
      ins->operands[i - 1] = val;
    }
  }
  return true;
}

// std::string_view InsSymbolize(const a64::Ins& ins,
//                              std::vector<std::string>* operands) {}

}  // namespace cwerg::a64
